### ISSUE TITLE
BIRTH — RN-20M: Ruse-Native Day Zero

### ISSUE THREAD

**Issue by bluegrouse92-lgtm** [N-A-pixel/Ruse-Public#2]
# BIRTH — RN-20M

This issue records the birth of the first **Ruse-Native model lineage**: RN-20M.

Until this milestone, Ruse's model experiments used pretrained/open-weight models such as Qwen, Phi and Llama-family artifacts as students, teachers or runtime probes. RN-20M changes that boundary. It is a small decoder-only transformer instantiated **from random initialization**, with no pretrained model weights inherited from Qwen, Phi, Llama, or another external model.

This is **Day Zero**, not a competence claim. RN-20M has not yet been trained to understand language or behave as Ruse. The milestone proves that Ruse now has its own reproducible model architecture and weight lineage that can travel through a real deployment lifecycle.

## What was born

RN-20M is a Hugging Face Llama-compatible decoder-only causal language model with:

```text
layers:                 10
hidden size:            384
FFN/intermediate:       1024
attention heads:        6
KV heads:               2
context:                1024
vocabulary:             16,000 (current smoke tokenizer)
embeddings:             tied
position encoding:      RoPE
normalization:          RMSNorm
activation/MLP:         SwiGLU
parameters:             21,880,704
```

Random initialization seed:

```text
20260921
```

The model lifecycle explicitly records:

```text
pretrained_weights_loaded = false
training_performed        = false
optimizer_created         = false
backward_performed        = false
```

No claim is made that random token generation represents intelligence or learned behaviour.

## Tokenizer

For the Stage 0 runtime proof, RN-20M uses a project-generated **smoke-only SentencePiece BPE tokenizer with byte fallback**.

```text
vocab size: 16,000
PAD:        0
BOS:        1
EOS:        2
UNK:        3
```

A literal LF token is present. The tokenizer was generated from a fixed generic fixture and did **not** use Ruse's 108-example behavioural corpus.

This tokenizer exists to prove the complete serialization/runtime lifecycle. It is **not the final Ruse tokenizer** and must not be represented as production-ready.

Known non-fatal runtime note: llama.cpp currently reports that EOS is not in its EOG set. This should be resolved as part of the eventual production-tokenizer work rather than hidden or ignored.

## Reproducible HF checkpoint

RN-20M was initialized with the frozen seed, packaged as a real self-contained Hugging Face checkpoint, and independently regenerated to verify deterministic artifacts.

Key Stage 0 hashes:

```text
model.safetensors SHA-256:
c47da781521c48b6640578063c05fc10d78593b8b120c2536fcb77e5546a54c8

GGUF SHA-256:
b5377e...a2e966

tokenizer.model SHA-256:
01f286...7905063
```

The complete hashes and environment provenance are recorded in the generated Stage 0 manifest.

## HF → GGUF → llama.cpp proof

The complete runtime lifecycle has now executed successfully:

```text
RN-20M architecture
→ deterministic random initialization
→ real SentencePiece tokenizer
→ Hugging Face save_pretrained checkpoint
→ llama.cpp convert_hf_to_gguf.py
→ F16 GGUF
→ native llama-cli
→ context initialization
→ random token generation
→ exit code 0
```

The official llama.cpp converter successfully detected the Llama architecture and SentencePiece tokenizer.

Generated F16 GGUF:

```text
size: 44,407,648 bytes
92 tensors loaded by llama.cpp
```

llama.cpp revision used for the proof:

```text
3479efd112b3910d2d008ad08fe4cb4895605903
```

The native Steam Deck build was produced outside the Ruse repository. Getting there required using the Android NDK Clang toolchain together with available local SDK/runtime components after the initial native GCC route failed. No llama.cpp build products were added to Ruse source.

## First runtime

`llama-cli` successfully loaded RN-20M's GGUF, recognized the 16K SentencePiece vocabulary, initialized a 128-token context and KV cache, generated eight tokens from the random model, and exited normally.

Measured Stage 0 runtime:

```text
peak RSS:              ~64.95 MB
elapsed:               ~0.084 s
prompt throughput:     ~2315.71 tok/s
generation throughput: ~448.11 tok/s
return code:           0
```

The generated output consisted of untrained/reserved-token nonsense. **That is the expected result.** Stage 0 tests deployment, not intelligence.

## Reproducibility

A second independent run using the same architecture, tokenizer procedure and seed was used to compare generated artifacts. The model checkpoint reproduced deterministically.

The Stage 0 evidence therefore gives Ruse a clean pre-learning control specimen: a known architecture, known seed, known tokenizer procedure and known random weight state against which later learning can be measured.

## Repository milestone

The Ruse-Native Stage 0 work was committed locally as:

```text
6dbc2d3 Add Ruse-Native RN-20M runtime lifecycle
```

Only the four intended Ruse-Native files were included in that commit. Existing Qwen/distillation worktree modifications remained separate and uncommitted.

At the time of this issue update, pushing that commit was blocked by an invalid local GitHub credential; the commit itself is preserved locally and should be pushed after re-authentication.

## Validation

Stage 0 completed with the existing repository gates intact:

```text
unit tests:                 13 passed
authored examples:          108 validated
quality gate:               passed
Python compile check:       passed
git diff --check:           passed
GGUF conversion:            passed
llama-cli runtime:          passed
```

No Qwen model or existing Qwen/distillation experiment was modified as part of RN-20M's runtime proof.

## What this proves

Stage 0 answers one question:

> **Can Ruse define, initialize, package, export and execute its own neural model from scratch?**

**Yes.**

It does **not** prove that RN-20M understands language, is intelligent, is Ruse-competent, or is ready for Wikipedia/research use.

The distinction is deliberate:

```text
Stage 0 — Can we build and run a Ruse-native model?       YES
Stage 1 — Can RN-20M measurably learn?                   UNKNOWN
Stage 2 — Can it acquire useful language competence?     UNKNOWN
Stage 3 — Can useful general pretraining scale?          UNKNOWN
Stage 4 — Can it acquire Ruse-specific behaviour?        UNKNOWN
Stage 5 — Can it pass independent research-readiness
          / Wikipedia-readiness evaluation?              UNKNOWN
```

## Why this matters

The earlier Qwen route represented a normal fine-tuning lineage: Ruse behaviour learned on top of somebody else's already-trained model weights.

RN-20M creates a different experimental lineage. Its 21,880,704 parameters begin at deterministic random initialization. If Stage 1 and later stages succeed, every learned capability in that lineage can be compared against this preserved Day Zero state.

The project can therefore ask a much cleaner developmental question:

> Can a small Ruse-native model progress from random initialization to measurable learning, language competence and eventually Ruse-specific behaviour using a grassroots £0 development path?

The answer after BIRTH is not yet yes or no.

For the first time, however, there is a real model with which to perform that experiment.

## Next milestone — Stage 1

Prepare a deliberately tiny **10K–100K-token controlled learning experiment**, beginning at the low end rather than spending compute unnecessarily.

The experiment should restart from the frozen RN-20M Day Zero state and answer only:

> **Can RN-20M measurably learn?**

Success must mean more than `training completed` or `loss changed`. Stage 1 should preserve a before/after control, record the learning curve, reload the trained checkpoint independently, and demonstrate acquisition of deliberately learnable patterns while keeping evaluation separate from training.

Do not begin general language pretraining or Ruse behavioural SFT until this basic learning capability has been demonstrated.

---

**BIRTH status:** RN-20M Stage 0 complete. 21,880,704 random parameters, reproducible HF checkpoint, F16 GGUF, native llama.cpp runtime, zero training tokens.

**Comment by bluegrouse92-lgtm** [comment]
## 23 September 2026 — authoritative OG parameter-count clarification

The preserved original RN-20M artifact was re-audited while resuming the historical Generation-1 lineage.

The original BIRTH record reported `21,880,704` parameters. The authoritative preserved architecture/checkpoint audit gives **21,865,344 parameters** for OG RN-20M.

The discrepancy is documented rather than silently rewriting the historical report.

Authoritative identifiers:

- Day Zero `model.safetensors`: `c47da781521c48b6640578063c05fc10d78593b8b120c2536fcb77e5546a54c8`
- protected C0 descendant: `cb2efe2c167f80c4d2e76477c79b195b6f4045925b3a6bd14330f3f5cca6fdbd`
- architecture: 10L / 384 hidden / 6 heads / 2 KV / FFN 1024 / context 1024 / vocab 16K / tied embeddings

The model identity and lineage do not change. Future records use **21,865,344** as the authoritative exact parameter count for OG RN-20M.
